src/Nethermind/Nethermind.Shutter/ShutterApi.cs

```diff
@@ -0,0 +1,162 @@
+            _cfg.ValidatorRegistryContractAddress!,
+            _cfg.ValidatorRegistryMessageVersion,
+            _txProcessingEnvFactory,
+            blockTree,
+            abiEncoder,
+            receiptFinder,
+            validatorsInfo,
+            Eon,
+            TxLoader,
+            Time,
+            logManager,
+            _slotLength,
+            BlockWaitCutoff);
+
+        TxSource = new ShutterTxSource(TxLoader, _cfg, Time, logManager);
+
+        KeyValidator = new ShutterKeyValidator(_cfg, Eon, logManager);
+
+        InitP2P(_cfg, logManager);
+        RegisterOnKeysValidated();
+    }
+
+    public void StartP2P(CancellationTokenSource? cancellationTokenSource = null)
+        => P2P!.Start(cancellationTokenSource);
+
+    public virtual void NewHeadBlockHandler(object? sender, BlockEventArgs e)
+    {
+        BlockHandler.OnNewHeadBlock(e.Block);
+    }
+
+    public ShutterBlockImprovementContextFactory GetBlockImprovementContextFactory(IBlockProducer blockProducer)
+    {
+        BlockImprovementContextFactory ??= new(
+            blockProducer,
+            TxSource,
+            _cfg,
+            Time,
+            _logManager,
+            _slotLength
+        );
+        return BlockImprovementContextFactory;
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        TxSource.Dispose();
+        BlockHandler.Dispose();
+        await (P2P?.DisposeAsync() ?? default);
+    }
+
+    protected void KeysReceivedHandler(object? sender, DecryptionKeys keys)
+    {
+        KeyValidator.OnDecryptionKeysReceived(keys);
+    }
+
+    protected virtual async void KeysValidatedHandler(object? sender, IShutterKeyValidator.ValidatedKeyArgs keys)
+    {
+        Metrics.TxPointer = keys.TxPointer;
+
+        // wait for latest block before loading transactions
+        Block? head = (await BlockHandler.WaitForBlockInSlot(keys.Slot - 1, new())) ?? _blockTree.Head;
+        BlockHeader? header = head?.Header;
+        BlockHeader parentHeader = header is not null
+            ? _blockTree.FindParentHeader(header, BlockTreeLookupOptions.None)!
+            : _blockTree.FindLatestHeader()!;
+        TxSource.LoadTransactions(head, parentHeader, keys);
+    }
+
+    protected virtual void InitP2P(IShutterConfig cfg, ILogManager logManager)
+    {
+        P2P = new(cfg, logManager);
+        P2P.KeysReceived += KeysReceivedHandler;
+    }
+
+    protected virtual void RegisterOnKeysValidated()
+    {
+        KeyValidator.KeysValidated += KeysValidatedHandler;
+    }
+
+    protected virtual IShutterEon InitEon()
+        => new ShutterEon(_blockTree, _txProcessingEnvFactory, _abiEncoder, _cfg, _logManager);
+
+    protected virtual ShutterTime InitTime(ISpecProvider specProvider, ITimestamper timestamper)
+    {
+        ulong genesisTimestamp = specProvider.ChainId == BlockchainIds.Chiado ? ChiadoSpecProvider.BeaconChainGenesisTimestamp : GnosisSpecProvider.BeaconChainGenesisTimestamp;
```

Comment:
this should be taken from available specProvider and not having `ifs` for different networks.